# Отчёт: базовые алгоритмы, главы 1.1-1.2 / Report: basic algorithms, chapters 1.1-1.2

Учебник / Textbook: Anany Levitin, *Introduction to the Design and Analysis of Algorithms*, 3rd ed.

Содержание / Contents:
1. Find max value in the array
2. Average of array's numbers
3. Ex. 1.1.4 - floor(sqrt(n))
4. Ex. 1.1.6 - gcd(31415, 14142)
5. Ex. 1.1.7 - proof of gcd(m, n) = gcd(n, m mod n)
6. Ex. 1.1.8 - Euclid's algorithm when m < n
7. Ex. 1.1.9 - min/max number of divisions for 1 <= m, n <= 10
8. Ex. 1.2.4 - roots of a quadratic equation
9. Ex. 1.2.5 - decimal to binary
10. Ex. 1.2.6 - ATM cash dispenser
11. Ex. 1.2.9 - improving MinDistance

В каждом пункте: задача, объяснение, псевдокод, код на Python, результат, эффективность.
Each item has: task, explanation, pseudocode, Python code, result, efficiency.

Базовая операция - операция, которая выполняется чаще всего; по её количеству оценивают скорость алгоритма.
Basic operation is the operation executed most often; counting it gives the running time.

## 1. Find max value in the array

**Задача:** найти наибольший элемент массива из n чисел.
**Task:** find the largest element of an array of n numbers.

**Объяснение:** берём первый элемент как текущий максимум, затем проходим по остальным и заменяем максимум, если встретили больший.
**Explanation:** take the first element as the current maximum, scan the rest, and replace the maximum whenever a bigger element appears.

Размер входа: n. Базовая операция: сравнение `A[i] > maxval`.

```
ALGORITHM MaxElement(A[0..n-1])
    maxval <- A[0]
    for i <- 1 to n-1 do
        if A[i] > maxval
            maxval <- A[i]
    return maxval
```

In [1]:
def max_element(A):
    maxval = A[0]                  # первый элемент - текущий максимум / first element is the current max
    for i in range(1, len(A)):     # проверяем элементы с индекса 1 / check elements from index 1
        if A[i] > maxval:          # базовая операция: сравнение / basic operation: comparison
            maxval = A[i]
    return maxval


print(max_element([62, 31, 84, 96, 19, 47]))   # 96
print(max_element([5]))                         # 5, цикл не выполняется / loop does not run
print(max_element([-10, -3, -50, -1]))          # -1, работает с отрицательными / works with negatives

96
5
-1


**Эффективность / Efficiency:** сравнение выполняется n - 1 раз, $C(n) = n - 1 \in \Theta(n)$. Лучший, худший и средний случаи совпадают.
The comparison runs n - 1 times; best, worst and average cases are the same.

## 2. Average of array's numbers

**Задача:** вычислить среднее арифметическое n чисел.
**Task:** compute the arithmetic mean of n numbers.

**Объяснение:** складываем все элементы в переменную `total` и делим сумму на n.
**Explanation:** add all elements into `total`, then divide by n.

Размер входа: n. Базовая операция: сложение.

```
ALGORITHM Average(A[0..n-1])
    sum <- 0
    for i <- 0 to n-1 do
        sum <- sum + A[i]
    return sum / n
```

In [2]:
def average(A):
    total = 0                      # накопитель суммы / accumulator
    for i in range(len(A)):
        total = total + A[i]       # базовая операция: сложение / basic operation: addition
    return total / len(A)          # одно деление в конце / one division at the end


print(average([62, 31, 84, 96, 19, 47]))        # 339 / 6 = 56.5
print(average([1, 1, 2, 3, 0, 1, 1, 0, 9]))     # 18 / 9 = 2.0

56.5
2.0


**Эффективность / Efficiency:** n сложений, $C(n) = n \in \Theta(n)$. Одно деление на порядок роста не влияет.
n additions; the single division does not change the order of growth.

## 3. Ex. 1.1.4 - floor(sqrt(n))

**Задача:** найти $\lfloor\sqrt{n}\rfloor$ для положительного целого n, используя только присваивание, сравнение и арифметику (без `sqrt`).
**Task:** find $\lfloor\sqrt{n}\rfloor$ for a positive integer n using only assignment, comparison and arithmetic (no `sqrt`).

**Объяснение:** увеличиваем t = 0, 1, 2, ... пока $(t+1)^2 \le n$. Когда условие нарушилось, t и есть ответ. Например, для n = 17: t доходит до 4, так как 5² = 25 > 17.
**Explanation:** increase t = 0, 1, 2, ... while $(t+1)^2 \le n$. When the condition fails, t is the answer. For n = 17, t reaches 4 because 5² = 25 > 17.

```
ALGORITHM FloorSqrt(n)
    t <- 0
    while (t + 1) * (t + 1) <= n do
        t <- t + 1
    return t
```

In [3]:
def floor_sqrt(n):
    t = 0
    while (t + 1) * (t + 1) <= n:  # можно ли сделать ещё шаг? / can we take one more step?
        t = t + 1
    return t


for n in [0, 1, 4, 15, 16, 17, 99, 100]:
    print(n, "->", floor_sqrt(n))

0 -> 0
1 -> 1
4 -> 2
15 -> 3
16 -> 4
17 -> 4
99 -> 9
100 -> 10


**Эффективность / Efficiency:** цикл выполняется около $\sqrt{n}$ раз, $C(n) \approx \sqrt{n} \in \Theta(\sqrt{n})$. Бинарный поиск (глава 4) даёт $\Theta(\log n)$.
The loop runs about $\sqrt{n}$ times. Binary search (chapter 4) gives $\Theta(\log n)$.

## 4. Ex. 1.1.6 - gcd(31415, 14142)

**Задача:** a) найти gcd(31415, 14142) алгоритмом Евклида; b) оценить, во сколько раз Евклид быстрее перебора последовательных целых чисел, начиная с min{m, n}.
**Task:** a) find gcd(31415, 14142) with Euclid's algorithm; b) estimate how many times faster it is than consecutive integer checking starting from min{m, n}.

**Объяснение:** Евклид использует равенство gcd(m, n) = gcd(n, m mod n) и повторяет его, пока остаток не станет 0. Перебор пробует t = min{m, n}, min{m, n} - 1, ... пока t не разделит оба числа.
**Explanation:** Euclid repeats gcd(m, n) = gcd(n, m mod n) until the remainder is 0. Consecutive checking tries t = min{m, n}, min{m, n} - 1, ... until t divides both numbers.

```
ALGORITHM Euclid(m, n)
    while n != 0 do
        r <- m mod n
        m <- n
        n <- r
    return m
```

In [4]:
def euclid_gcd(m, n):
    divisions = 0
    while n != 0:
        m, n = n, m % n            # новое m = старое n, новое n = остаток / new m = old n, new n = remainder
        divisions += 1
    return m, divisions


def consecutive_integer_checking_gcd(m, n):
    t = min(m, n)
    iterations = 0
    while True:
        iterations += 1
        if m % t == 0 and n % t == 0:   # t делит оба числа / t divides both numbers
            return t, iterations
        t -= 1


m, n = 31415, 14142
gcd_e, div_e = euclid_gcd(m, n)
gcd_c, iters_c = consecutive_integer_checking_gcd(m, n)

print(f"gcd({m}, {n}) = {gcd_e}")
print(f"Euclid: {div_e} divisions / Евклид: {div_e} делений")
print(f"Checking: {iters_c} iterations / Перебор: {iters_c} итераций")
print(f"Euclid is faster about {iters_c / div_e:.0f} times / Евклид быстрее примерно в {iters_c / div_e:.0f} раз")

gcd(31415, 14142) = 1
Euclid: 10 divisions / Евклид: 10 делений
Checking: 14142 iterations / Перебор: 14142 итераций
Euclid is faster about 1414 times / Евклид быстрее примерно в 1414 раз


**Вывод / Conclusion:** gcd = 1. Евклиду нужно 10 делений, перебору 14142 итерации, разница около 1414 раз. Число делений у Евклида растёт логарифмически, у перебора - до min(m, n).
gcd = 1. Euclid needs 10 divisions, checking needs 14142 iterations, about 1414 times more. Euclid grows logarithmically, checking grows up to min(m, n).

## 5. Ex. 1.1.7 - proof of gcd(m, n) = gcd(n, m mod n)

**Задача:** доказать, что gcd(m, n) = gcd(n, m mod n).
**Task:** prove that gcd(m, n) = gcd(n, m mod n).

**Объяснение:** показываем, что у пар (m, n) и (n, r) одинаковые множества общих делителей, поэтому одинаков и наибольший из них.
**Explanation:** we show that the pairs (m, n) and (n, r) have the same set of common divisors, so their greatest elements are equal.

Пусть $r = m \bmod n$, тогда $m = qn + r$.
Let $r = m \bmod n$, so $m = qn + r$.

- Если d делит m и n, то d делит $r = m - qn$.
  If d divides m and n, then d divides $r = m - qn$.
- Если d делит n и r, то d делит $m = qn + r$.
  If d divides n and r, then d divides $m = qn + r$.

Значит, общие делители совпадают, и $\gcd(m, n) = \gcd(n, r)$. $\blacksquare$
So the common divisors coincide and $\gcd(m, n) = \gcd(n, r)$.

Ниже - численная проверка на случайных парах (это не доказательство, а проверка формулы).
Below is a numeric check on random pairs (not a proof, just a test of the formula).

In [5]:
import math
import random

random.seed(42)

for _ in range(10):
    m = random.randint(1, 10_000)
    n = random.randint(1, 10_000)
    left = math.gcd(m, n)
    right = math.gcd(n, m % n)
    print(f"gcd({m}, {n}) = {left}   gcd({n}, {m % n}) = {right}   {'OK' if left == right else 'ERROR'}")

gcd(1825, 410) = 5   gcd(410, 185) = 5   OK
gcd(4507, 4013) = 1   gcd(4013, 494) = 1   OK
gcd(3658, 2287) = 1   gcd(2287, 1371) = 1   OK
gcd(1680, 8936) = 8   gcd(8936, 1680) = 8   OK
gcd(1425, 9675) = 75   gcd(9675, 1425) = 75   OK
gcd(6913, 521) = 1   gcd(521, 140) = 1   OK
gcd(489, 1536) = 3   gcd(1536, 489) = 3   OK
gcd(3583, 3812) = 1   gcd(3812, 3583) = 1   OK
gcd(8280, 9864) = 72   gcd(9864, 8280) = 72   OK
gcd(435, 9196) = 1   gcd(9196, 435) = 1   OK


## 6. Ex. 1.1.8 - Euclid's algorithm when m < n

**Задача:** что делает алгоритм Евклида, если первое число меньше второго, и сколько раз это может случиться?
**Task:** what does Euclid's algorithm do when the first number is smaller than the second, and how many times can it happen?

**Объяснение:** если m < n, то m mod n = m, поэтому Euclid(m, n) превращается в Euclid(n, m). Первый шаг просто меняет числа местами. После этого первое число больше второго, и дальше остаток всегда меньше делителя. Значит, такая ситуация возможна только один раз, на первом шаге.
**Explanation:** if m < n then m mod n = m, so Euclid(m, n) becomes Euclid(n, m). The first step only swaps the numbers. After that the first number is larger, and the remainder is always smaller than the divisor. So this can happen only once, at the first step.

In [6]:
def euclid_trace(m, n):
    while n != 0:
        print(f"m={m}, n={n}, m < n: {m < n}")
        m, n = n, m % n


euclid_trace(14142, 31415)
print()
euclid_trace(100, 999)

m=14142, n=31415, m < n: True
m=31415, n=14142, m < n: False
m=14142, n=3131, m < n: False
m=3131, n=1618, m < n: False
m=1618, n=1513, m < n: False
m=1513, n=105, m < n: False
m=105, n=43, m < n: False
m=43, n=19, m < n: False
m=19, n=5, m < n: False
m=5, n=4, m < n: False
m=4, n=1, m < n: False

m=100, n=999, m < n: True
m=999, n=100, m < n: False
m=100, n=99, m < n: False
m=99, n=1, m < n: False


**Результат / Result:** `m < n` равно True только в первой строке каждой трассировки, во всех остальных строках False.
`m < n` is True only in the first line of each trace and False in all other lines.

## 7. Ex. 1.1.9 - min and max number of divisions for 1 <= m, n <= 10

**Задача:** найти минимальное и максимальное число делений алгоритма Евклида среди всех пар $1 \le m, n \le 10$.
**Task:** find the minimum and maximum number of divisions of Euclid's algorithm over all pairs $1 \le m, n \le 10$.

**Объяснение:** перебираем все 100 пар, для каждой считаем число делений и сохраняем в словарь `results`. Затем берём минимум и максимум и выводим пары, на которых они достигаются. В конце печатаем таблицу 10x10.
**Explanation:** we go through all 100 pairs, count the divisions for each one and store them in the dictionary `results`. Then we take the minimum and maximum and print the pairs where they occur. At the end we print the 10x10 table.

In [7]:
def euclid_division_count(m, n):
    count = 0
    while n != 0:
        m, n = n, m % n
        count += 1
    return count


results = {}
for m in range(1, 11):
    for n in range(1, 11):
        results[(m, n)] = euclid_division_count(m, n)

min_div = min(results.values())
max_div = max(results.values())

print("Min:", min_div, "pairs:", [p for p, c in results.items() if c == min_div])
print("Max:", max_div, "pairs:", [p for p, c in results.items() if c == max_div])

print()
print("      " + " ".join(f"n={n:<2d}" for n in range(1, 11)))
for m in range(1, 11):
    row = " ".join(f"{results[(m, n)]:4d}" for n in range(1, 11))
    print(f"m={m:<2d} {row}")

Min: 1 pairs: [(1, 1), (2, 1), (2, 2), (3, 1), (3, 3), (4, 1), (4, 2), (4, 4), (5, 1), (5, 5), (6, 1), (6, 2), (6, 3), (6, 6), (7, 1), (7, 7), (8, 1), (8, 2), (8, 4), (8, 8), (9, 1), (9, 3), (9, 9), (10, 1), (10, 2), (10, 5), (10, 10)]
Max: 5 pairs: [(5, 8)]

      n=1  n=2  n=3  n=4  n=5  n=6  n=7  n=8  n=9  n=10
m=1     1    2    2    2    2    2    2    2    2    2
m=2     1    1    3    2    3    2    3    2    3    2
m=3     1    2    1    3    4    2    3    4    2    3
m=4     1    1    2    1    3    3    4    2    3    3
m=5     1    2    3    2    1    3    4    5    4    2
m=6     1    1    1    2    2    1    3    3    3    4
m=7     1    2    2    3    3    2    1    3    4    4
m=8     1    1    3    1    4    2    2    1    3    3
m=9     1    2    1    2    3    2    3    2    1    3
m=10    1    1    2    2    1    3    3    2    2    1


**Результат / Result:**
- Минимум 1 деление, когда n делит m (например, n = 1 или m = n). / Minimum is 1 division, when n divides m (for example n = 1 or m = n).
- Максимум 5 делений только на паре (5, 8). Пара (8, 5) даёт 4 деления, потому что первый шаг не нужен. Это соседние числа Фибоначчи, худший случай для Евклида (теорема Ламе).
  Maximum is 5 divisions, only for the pair (5, 8). The pair (8, 5) gives 4 because the swapping step is not needed. These are neighbouring Fibonacci numbers, the worst case for Euclid (Lamé's theorem).

## 8. Ex. 1.2.4 - roots of ax^2 + bx + c = 0

**Задача:** найти действительные корни уравнения $ax^2 + bx + c = 0$ для произвольных вещественных a, b, c.
**Task:** find the real roots of $ax^2 + bx + c = 0$ for arbitrary real a, b, c.

**Объяснение:** если a = 0, уравнение линейное или вырожденное. Иначе считаем дискриминант $D = b^2 - 4ac$: при D > 0 два корня, при D = 0 один корень, при D < 0 действительных корней нет.
**Explanation:** if a = 0 the equation is linear or degenerate. Otherwise compute $D = b^2 - 4ac$: two roots if D > 0, one root if D = 0, no real roots if D < 0.

```
ALGORITHM QuadraticRoots(a, b, c)
    if a = 0
        if b = 0
            if c = 0: return "infinitely many solutions"
            else:     return "no solutions"
        else
            return -c / b
    D <- b * b - 4 * a * c
    if D > 0
        return (-b + sqrt(D)) / (2a), (-b - sqrt(D)) / (2a)
    else if D = 0
        return -b / (2a)
    else
        return "no real roots"
```

In [8]:
import math


def quadratic_roots(a, b, c):
    if a == 0:                                 # вырожденный случай / degenerate case
        if b == 0:
            return "infinitely many solutions" if c == 0 else "no solutions"
        return -c / b                          # линейное уравнение bx + c = 0 / linear equation
    D = b * b - 4 * a * c                      # дискриминант / discriminant
    if D > 0:
        x1 = (-b + math.sqrt(D)) / (2 * a)
        x2 = (-b - math.sqrt(D)) / (2 * a)
        return x1, x2
    elif D == 0:
        return -b / (2 * a)
    else:
        return "no real roots"


print(quadratic_roots(1, -3, 2))    # x^2 - 3x + 2: roots 2 and 1
print(quadratic_roots(2, -7, 3))    # 2x^2 - 7x + 3: roots 3 and 0.5
print(quadratic_roots(1, 2, 1))     # one root -1
print(quadratic_roots(1, 0, 1))     # no real roots
print(quadratic_roots(0, 2, 1))     # linear: -0.5
print(quadratic_roots(0, 0, 0))     # infinitely many
print(quadratic_roots(0, 0, 5))     # no solutions

(2.0, 1.0)
(3.0, 0.5)
-1.0
no real roots
-0.5
infinitely many solutions
no solutions


**Проверка / Check:** по теореме Виета для $x^2 - 3x + 2$ сумма корней 3 = -b/a и произведение 2 = c/a; для $2x^2 - 7x + 3$ сумма 3.5 = 7/2 и произведение 1.5 = 3/2.
By Vieta's formulas, for $x^2 - 3x + 2$ the sum is 3 = -b/a and the product is 2 = c/a; for $2x^2 - 7x + 3$ the sum is 3.5 = 7/2 and the product is 1.5 = 3/2.

## 9. Ex. 1.2.5 - decimal to binary

**Задача:** описать алгоритм перевода положительного десятичного числа в двоичное (словами и псевдокодом).
**Task:** describe the algorithm that converts a positive decimal number to binary (in words and in pseudocode).

**Объяснение (словами):** делим n на 2 и записываем остаток (0 или 1). Заменяем n на целую часть частного. Повторяем, пока n не станет 0. Остатки, прочитанные в обратном порядке, дают двоичную запись. Пример: 13 даёт остатки 1, 0, 1, 1, значит 13 = 1101.
**Explanation (in words):** divide n by 2 and write down the remainder (0 or 1). Replace n with the integer part of the quotient. Repeat until n is 0. The remainders read in reverse order are the binary form. Example: 13 gives remainders 1, 0, 1, 1, so 13 = 1101.

```
ALGORITHM DecimalToBinary(n)
    digits <- empty list
    while n > 0 do
        append (n mod 2) to digits
        n <- floor(n / 2)
    reverse(digits)
    return digits
```

In [9]:
def decimal_to_binary(n):
    digits = []
    while n > 0:
        digits.append(n % 2)       # остаток - очередная цифра, от младшей к старшей / remainder is the next digit, lowest first
        n = n // 2                 # целочисленное деление / integer division
    digits.reverse()               # переворачиваем: старшая цифра первой / reverse: highest digit first
    return "".join(map(str, digits))


for n in [1, 2, 5, 10, 13, 25]:
    print(n, "->", decimal_to_binary(n))

b = decimal_to_binary(1024)
print(1024, "->", b, "| digits / разрядов:", len(b))

1 -> 1
2 -> 10
5 -> 101
10 -> 1010
13 -> 1101
25 -> 11001
1024 -> 10000000000 | digits / разрядов: 11


**Эффективность / Efficiency:** число делений равно числу двоичных разрядов, $C(n) = \lfloor \log_2 n \rfloor + 1 \in \Theta(\log n)$. Для n = 1024 нужно 11 делений.
The number of divisions equals the number of binary digits. For n = 1024 it is 11 divisions.

## 10. Ex. 1.2.6 - ATM cash dispenser

**Задача:** описать алгоритм, по которому банкомат выдаёт запрошенную сумму.
**Task:** describe the algorithm an ATM uses to dispense the requested amount.

**Объяснение:** жадный алгоритм. Номиналы идут по убыванию. Для каждого номинала берём максимально возможное число купюр, уменьшаем остаток и переходим к следующему номиналу. Если в конце остаток не 0, сумму выдать нельзя. Жадный подход оптимален не для любого набора номиналов, но для обычных банкнотных систем работает.
**Explanation:** a greedy algorithm. Denominations go from largest to smallest. For each one take as many bills as possible, reduce the remainder and move on. If the remainder is not 0 at the end, the amount cannot be paid out. Greedy is not optimal for every set of denominations, but it works for ordinary banknote systems.

```
ALGORITHM DispenseCash(amount, denominations[1..k])
    result <- empty dictionary
    remaining <- amount
    for i <- 1 to k do
        count <- floor(remaining / denominations[i])
        if count > 0
            result[denominations[i]] <- count
            remaining <- remaining - count * denominations[i]
    if remaining != 0
        return "cannot dispense this amount"
    return result
```

In [10]:
def dispense_cash(amount, denominations=(10000, 5000, 2000, 1000, 500, 200, 100)):
    result = {}
    remaining = amount
    for bill in denominations:               # от крупных к мелким / from large to small
        count = remaining // bill            # сколько купюр помещается / how many bills fit
        if count > 0:
            result[bill] = count
            remaining -= count * bill
    if remaining != 0:                       # остаток не выдать / remainder cannot be paid
        return "cannot dispense this amount"
    return result


print(dispense_cash(37600))
print(dispense_cash(12300))
print(dispense_cash(500))
print(dispense_cash(999))                    # не кратно 100 / not a multiple of 100

{10000: 3, 5000: 1, 2000: 1, 500: 1, 100: 1}
{10000: 1, 2000: 1, 200: 1, 100: 1}
{500: 1}
cannot dispense this amount


**Результат / Result:** 37600 = 3x10000 + 5000 + 2000 + 500 + 100. Сумма 999 не кратна минимальному номиналу 100, поэтому банкомат сообщает, что выдать её нельзя.
The amount 999 is not a multiple of the smallest bill 100, so the ATM reports that it cannot be paid out.

## 11. Ex. 1.2.9 - improving MinDistance

**Задача:** улучшить алгоритм поиска минимального расстояния между двумя элементами массива.
**Task:** improve the algorithm that finds the minimum distance between two elements of an array.

**Исходный алгоритм:**
```
ALGORITHM MinDistance(A[0..n-1])
    dmin <- infinity
    for i <- 0 to n-1 do
        for j <- 0 to n-1 do
            if i != j and |A[i] - A[j]| < dmin
                dmin <- |A[i] - A[j]|
    return dmin
```

**Объяснение улучшений / Explanation of improvements:**
1. Пара (i, j) и пара (j, i) дают одно расстояние, поэтому достаточно j > i. Сравнений становится вдвое меньше (n(n-1)/2), порядок остаётся $\Theta(n^2)$.
   The pairs (i, j) and (j, i) give the same distance, so j > i is enough. Comparisons are halved (n(n-1)/2), the order stays $\Theta(n^2)$.
2. Если массив отсортировать, ближайшие элементы окажутся соседями, и хватит одного прохода по соседним парам. Итого $\Theta(n \log n)$ (сортировка) и $\Theta(n)$ (проход).
   After sorting, the closest elements are neighbours, so one pass over adjacent pairs is enough. Total is $\Theta(n \log n)$ for sorting plus $\Theta(n)$ for the pass.

In [11]:
def min_distance_original(A):
    n = len(A)
    dmin = float('inf')
    for i in range(n):
        for j in range(n):
            if i != j and abs(A[i] - A[j]) < dmin:
                dmin = abs(A[i] - A[j])
    return dmin


def min_distance_improved(A):
    n = len(A)
    dmin = float('inf')
    for i in range(n - 1):
        for j in range(i + 1, n):            # каждая пара один раз / each pair once
            if abs(A[i] - A[j]) < dmin:
                dmin = abs(A[i] - A[j])
    return dmin


def min_distance_sorted(A):
    B = sorted(A)                            # отсортированная копия / sorted copy
    dmin = float('inf')
    for i in range(len(B) - 1):
        diff = B[i + 1] - B[i]               # abs не нужен, B отсортирован / no abs needed, B is sorted
        if diff < dmin:
            dmin = diff
    return dmin


arr = [1, 5, 3, 19, 18, 25]
print(min_distance_original(arr), min_distance_improved(arr), min_distance_sorted(arr))   # 1 1 1

1 1 1


Сравним время работы на массивах разного размера. `time.perf_counter()` возвращает точное время, разность двух вызовов - время работы.
Now the running time on arrays of different sizes. `time.perf_counter()` gives a precise timestamp; the difference of two calls is the running time.

In [12]:
import time
import random

random.seed(2)

for n in [100, 500, 1000, 2000, 4000]:
    arr = [random.randint(-100000, 100000) for _ in range(n)]

    t0 = time.perf_counter()
    d1 = min_distance_original(arr)
    t1 = time.perf_counter()
    d2 = min_distance_improved(arr)
    t2 = time.perf_counter()
    d3 = min_distance_sorted(arr)
    t3 = time.perf_counter()

    assert d1 == d2 == d3                    # все версии дают один ответ / all versions agree
    print(f"n={n:5d}  original={t1 - t0:.3f}s  improved={t2 - t1:.3f}s  sorted={t3 - t2:.4f}s")

n=  100  original=0.001s  improved=0.000s  sorted=0.0000s
n=  500  original=0.017s  improved=0.010s  sorted=0.0001s
n= 1000  original=0.060s  improved=0.024s  sorted=0.0002s


n= 2000  original=0.279s  improved=0.097s  sorted=0.0004s


n= 4000  original=0.947s  improved=0.432s  sorted=0.0007s


**Итог / Summary:**

| Версия / Version | Порядок / Order | Комментарий / Comment |
|---|---|---|
| original | $\Theta(n^2)$, около n² сравнений | каждая пара дважды / each pair twice |
| improved | $\Theta(n^2)$, около n²/2 сравнений | вдвое быстрее, тот же порядок / twice faster, same order |
| sorted | $\Theta(n \log n)$ | другой порядок роста / better order of growth |

При росте n время original и improved растёт квадратично, а sorted почти не меняется.
As n grows, original and improved grow quadratically, while sorted barely changes.